Feature/send

FredrikOseberg opened the pull request:
I will complete the survey below once this is ready to be merged. As this feature will be fairly large and critical to Neon I'm putting this up to receive early feedback on the PR. So far I've completed the initial steps in setting up the UI and functionality for adding / editing send rows with information that is reflected in state.

**What current issue(s) from Trello/Github does this address?**

**What problem does this PR solve?**

**How did you solve this problem?**

**How did you make sure your solution works?**

**Are there any special changes in the code that we should be aware of?**

**Is there anything else we should know?**

- [ ] Unit tests written?

dvdschwrtz commented:
great work so far! Code looks good

comountainclimber commented on app/components/Button/Button.jsx:
```diff
@@ -10,25 +10,32 @@ type Props = {
   renderIcon: ?Function,
   primary: ?boolean,
   type: ?string,
+  displayButtonIcon?: boolean,
```
By excluding the prop `renderIcon` Im pretty sure this component will render a button without the icon so displayButtonIcon might be unnecessary unless Im mistaken

FredrikOseberg replied:
I thought so too, however as far as I can tell it will render an empty div thats still takes up space and skews the text in the button: 
```
  renderIcon = () => {
    const { renderIcon } = this.props

    if (renderIcon) {
      return (
        <span className={classNames(styles.icon, this.getIconStyle())}>
          {renderIcon()}
        </span>
      )
    }
    return <div className={styles.icon} />
  }
```

comountainclimber commented on app/components/Send/SendAmountsPanel/index.jsx:
```diff
@@ -0,0 +1,29 @@
+import SendAmountsInfoBox from './SendAmountsInfoBox'
+
+import styles from './SendAmountsPanel.scss'
+
```
Lets add proptype definitions here

FredrikOseberg replied:
Yep. Noted.

comountainclimber commented on app/components/Send/SendAmountsPanel/index.jsx:
```diff
@@ -0,0 +1,29 @@
+class SendPageHeader extends Component {
+  render() {
+    return (
+      <section className={styles.sendAmountsPanel}>
```
Might be cool to have this component scroll horizontally if user has many tokens
<img width="901" alt="screen shot 2018-07-07 at 12 03 52 am" src="https://user-images.githubusercontent.com/13072035/42407410-55f7427c-8179-11e8-822c-d1222ad68380.png">

FredrikOseberg replied:
Agree. What would be the best way? Stacked on top of each other and infinite length, or define a max length and set overflow: scroll ?

FredrikOseberg replied:
Yes, that would work, I'll implement and we can review if it's clear enough for the user. Perhaps I'll add some sort of indicator when the total amount of tokens exceed the space in the bar. What do you think?

comountainclimber commented on package.json:
```diff
@@ -135,7 +135,7 @@
     "cross-env": "5.1.1",
     "css-loader": "0.28.7",
-    "electron": "1.8.4",
+    "electron": "1.7.11",
```
Whats this about?

FredrikOseberg replied:
I ran into a problem when cloning the application and running NPM install. NPM installed electron 1.8.4 when the readme stated that 1.7.11 should be used. Once I updated the electron version the application ran without issues. Are you running 1.8.4 without problems?

zsiglin replied:
I haven't run into any problems thus far, but @FredrikOseberg is correct; the README needs to be updated.

FredrikOseberg replied:
<img width="1050" alt="screen shot 2018-07-08 at 13 53 03" src="https://user-images.githubusercontent.com/16081982/42419516-b691d974-82b6-11e8-8ebc-0f2654de539c.png">

comountainclimber replied:
Please revert this change 👆

comountainclimber commented:
I agree with @dvdschwrtz this looks awesome man! Impressed with how far along you are and glad you opened up this PR before digging into the other components.

Would ideally like to get this merged in and have QR NEP9, and priority functionality can be in a separate PR. Noticed you still havent gotten some of the actions wired up including loading account balances and the ability to actually send assets. Let me know if you need any help with that this looks great so far!

FredrikOseberg commented:
Thanks for the feedback! The next step will be to integrate the UI with actual data, perform validation and actually send assets. I'll defer priority and QR code sending to a separate PR and focus purely on sending asset as the core functionality of this PR.

FredrikOseberg commented on app/components/Send/SendAmountsPanel/SendAmountsInfoBox/index.jsx:
```diff
@@ -0,0 +1,42 @@
+      </p>
+      <p className={styles.remainingAssetValue}>
+        {fiatCurrencySymbol}
+        {assetPrice * remainingAmount}
```
Thanks @evgenyboxer. I'll implement these changes as requested.

comountainclimber commented:
@FredrikOseberg fantastic job so far man. Just so you are aware if you run `jest --updateSnapshot` your CI tests will pass 👍

Please let me know if you are blocked or need assistance on specific remaining tasks

FredrikOseberg commented:
@comountainclimber @dvdschwrtz @evgenyboxer @zsiglin 

This PR is very close to being finished. Here's a brief overview over what's been done and things I have changed: 

1. Implemented Send functionality according to the new designs, filling in the blanks where there were no design specifications available. Current flow:
* If no assets are available, the user will be informed he has no assets to send.
* If assets are available, the user will be shown a screen like depicted in the design with fields for asset, amount, address and note. Users can add or remove rows, but not more than 5 and not less than 1.
* If users fill out the form and click send, they will be shown a screen like depicted in the designs where they are asked to look over the details and confirm the send details to avoid loss of funds.
* If sending is successful, users will be shown a screen that informs them of this and displays a list of transactions. In this list, the users can see the amount, address and note. And they can add a user to their contact list, copy the address or view the transaction on Neoscan.
* If sending fails, the user will be shown a failure screen with the error message.

**Major Changes needing review**
1. Edited sendTransaction function to be a promise to allow for proper catching and subsequent UI rendering. Also made sendTransaction return the result with the txid.
2. Changed the design of modal headers to be more in line with design. It no longer displays the modal title, but displays the logo of Neon and the title: "NEON". **Reasoning**: The modals representing current functionality will be swapped out, and the this is more in line with what is depicted in the design.
3. Changed blacklist to be hardcoded into Neon source code. **Reasoning**: With the current implementation of the blacklist (getting it via ajax call from github gists), we are one bad ajax call away from allowing users to send to known phishing addresses. This is a list that needs to be manually hardcoded somewhere, and it may as well be in the Neon wallet. Thoughts?
4. Changed the design of Add Contact to be more in line with design.

**Smaller changes**
* Smaller changes to input fields. Added error styles and the ability to accept error text for several input fields.
* Added two utilities: pluralize and createFormattedDate

**Testing**
The feature has been thoroughly tested. You can view the tests in Send.test.js.

Remaining: 
* Clean up and remove old Send functionality
* Restore Electron version to 1.8.4 (as you know I was unable to get it working on 1.8.4)

I'd love to get some more eyes on the code. I've been working on this for a while, so there is likely some things I'm missing.

comountainclimber commented on app/util/blacklist.js:
```diff
@@ -0,0 +1,35 @@
+export default [
+  'ARU4Sw9yyqgfjxfqF1TNwWHHFvLbAVdTj1',
```
Can we keep the original call but fall back to this if it fails?

comountainclimber commented on app/styles/variables.scss:
```diff
@@ -28,6 +29,12 @@ $top-margin: 20px;
+$white: #fff;
+$dark-grey: #515151;
+$light-grey: #f2f2f2;
+$medium-grey: #8a8383;
```
👍

comountainclimber commented on app/containers/WalletManager/WalletManager.scss:
```diff
@@ -6,14 +8,14 @@
   align-items: center;
-  background: #f2f2f2;
-  color: #515151;
+  background: $light-grey;
```
This will make our "theme" switch feature much easier thanks

comountainclimber commented:
@FredrikOseberg This is really great work! Nice job building up our test coverage as well 💪 

Although still reviewing a couple things to note:
- Please restore electron back to original version in `design-v2` the bug you were experiencing has been resolved
- "Changed blacklist to be hardcoded into Neon source code." Lets leave the ajax call in place but fall back to hardcoded values if it fails that way we do not have to release every time we add a scam address
- Please remove the old send functionality entirely.
- What you have stubbed out in terms of design for a wallet with no assets looks fine for now


When I try and access the send functionality locally on a wallet with assets the app throws an error:
<img width="693" alt="screen shot 2018-07-16 at 10 45 46 pm" src="https://user-images.githubusercontent.com/13072035/42796968-2e9bbfb8-894a-11e8-9823-b618893a894a.png">

So its hard for me to review or do additional testing. Please let me know when this is resolved. Thanks again @FredrikOseberg nice work man!

comountainclimber commented:
Super close on being able to merge this in addition to my comments above have just two more concerns

1. Please use currency set by users from settings. Also currencies should be appropriately rounded look in `formatters.js`

2. Lets go ahead and drop the notes column as its a _nice to have_ and not necessary to implement at this stage

comountainclimber commented on app/components/Send/DisplayInput/index.jsx:
```diff
@@ -0,0 +1,18 @@
+// @flow
+
```
No line break necessary after Flow declaration.

comountainclimber commented on app/components/Send/SendPanel/ConfirmSend/index.jsx:
```diff
@@ -0,0 +1,44 @@
+// @flow
+
```
No line break necessary after flow declaration

comountainclimber commented on app/components/Send/SendPanel/SendError/index.jsx:
```diff
@@ -0,0 +1,46 @@
+// @flow
+
```
Same as comments above

comountainclimber commented on app/components/Send/SendPanel/SendRecipientList/SendRecipientListItem/index.jsx:
```diff
@@ -0,0 +1,170 @@
+// @flow
+
```
No line break necessary after flow declaration

comountainclimber commented on app/components/Send/SendPanel/SendRecipientList/index.jsx:
```diff
@@ -0,0 +1,77 @@
+// @flow
+
```
No line break necessary after flow declaration

comountainclimber commented on app/components/Send/SendPanel/SendSuccess/index.jsx:
```diff
@@ -0,0 +1,58 @@
+// @flow
+
```
No line break necessary after flow declaration

comountainclimber commented on app/components/Send/SendPanel/index.jsx:
```diff
@@ -0,0 +1,138 @@
+// @flow
+
```
No line break necessary after flow declaration

comountainclimber commented on app/containers/Send/index.js:
```diff
@@ -0,0 +1,69 @@
+import { compose } from 'recompose'
```
This file is missing flow declaration (see line 48)

comountainclimber commented on app/util/createFormattedDate.js:
```diff
@@ -0,0 +1,36 @@
+export const createFormattedDate = () => {
```
While this looks fine to be is there a reason we arent just using moment js here?

comountainclimber commented on app/util/pluralize.js:
```diff
@@ -0,0 +1,3 @@
+// @flow
+export const pluralize = (word: string, items: number) =>
```
👍 nice

comountainclimber commented on __tests__/components/Send.test.js:
```diff
@@ -0,0 +1,216 @@
+    const errors = wrapper.instance().state.sendRowDetails[0].errors
+    expect(errors.amount).toBe('You cannot send fractional amounts of NEO.')
+  })
+
```
These tests look great man! Can we add one more verifying that an error state occurs user attempts to send assets to themself

comountainclimber commented on app/components/Contacts/ContactForm/ContactForm.jsx:
```diff
@@ -2,21 +2,35 @@
   onSubmit: Function
 }
 
 export default class ContactForm extends React.Component<Props> {
```
Please add type checking for State `<Props,State>` AND
```
type State = {
  nameError: string,
  addressError: string
}
```

comountainclimber commented on app/components/Contacts/ContactForm/ContactForm.jsx:
```diff
@@ -27,46 +41,168 @@ export default class ContactForm extends React.Component<Props> {
+  componentWillMount() {
+    const { newAddress, setAddress, address } = this.props
+
+    this.setState({ ownAddress: address })
```
Cant the state object just be initialized with this value as opposed to being set in componentWillMount?

comountainclimber commented on app/components/Contacts/ContactForm/ContactForm.jsx:
```diff
@@ -27,46 +41,168 @@ export default class ContactForm extends React.Component<Props> {
+    return validName && validAddress
+  }
+
+  validateName = name => {
```
Please add type annotations here

comountainclimber commented on app/components/Contacts/ContactForm/ContactForm.jsx:
```diff
@@ -27,46 +41,168 @@ export default class ContactForm extends React.Component<Props> {
+    }
+  }
+
+  validate = (name, address) => {
```
Please add type annotations here

comountainclimber commented on app/components/Contacts/ContactForm/ContactForm.jsx:
```diff
@@ -27,46 +41,168 @@ export default class ContactForm extends React.Component<Props> {
+    return true
+  }
+
+  validateAddress = address => {
```
Please add type annotations here

comountainclimber commented on app/components/Contacts/ContactForm/ContactForm.jsx:
```diff
@@ -27,46 +41,168 @@ export default class ContactForm extends React.Component<Props> {
+    return true
+  }
+
+  clearErrors = name => {
```
Please add type annotations here

comountainclimber commented on app/components/Contacts/ContactForm/ContactForm.jsx:
```diff
@@ -27,46 +41,168 @@ export default class ContactForm extends React.Component<Props> {
     this.props.setName(event.target.value)
   }
 
   handleChangeAddress = (event: Object) => {
```
`SyntheticMouseEvent<*>`

comountainclimber commented on app/components/Contacts/ContactForm/ContactForm.jsx:
```diff
@@ -27,46 +41,168 @@ export default class ContactForm extends React.Component<Props> {
+    }
+  }
+
   handleChangeName = (event: Object) => {
```
`SyntheticMouseEvent<*>`

comountainclimber commented on app/components/Send/SendAmountsPanel/index.jsx:
```diff
@@ -0,0 +1,30 @@
+        remainingAmount={dataset.currentBalance}
+        totalBalanceWorth={dataset.totalBalanceWorth}
+        remainingBalanceWorth={dataset.remainingBalanceWorth}
+        fiatCurrencySymbol="$"
```
Look in constants.CURRENCIES for correct mapping 👍

comountainclimber commented on app/containers/Send/Send.jsx:
```diff
@@ -0,0 +1,362 @@
+  contacts: Object
+}
+
+export default class Send extends React.Component<Props> {
```
Lets document state here with type annotations as well...

comountainclimber commented on app/containers/Send/Send.jsx:
```diff
@@ -0,0 +1,362 @@
+      )
+    }
+
+    return assets.filter(asset => !!prices[asset]).map(asset => {
```
```
    return (assets.filter((asset: string): Boolean => !!prices[asset]).map(
      (asset: string): Object => {
        const { balance } = sendableAssets[asset]
        const currentBalance = minusNumber(
          balance,
          this.calculateRowAmounts(asset)
        )
        const price = prices[asset]

        const totalBalanceWorth = multiplyNumber(balance, price)
        const remainingBalanceWorth = multiplyNumber(currentBalance, price)

        return {
          symbol: asset,
          totalBalance: balance,
          price,
          currentBalance,
          totalBalanceWorth,
          remainingBalanceWorth
        }
      }
    ): Array<*>)
```
